### PR TITLE
cmaketoolchain extensibility

### PR DESCRIPTION
**Pull request by memsharded**
Docs for: https://github.com/conan-io/conan/pull/8749

### CODE REVIEW

File: reference/conanfile/tools/cmake.rst [hunk 1]
```diff
@@ -192,11 +184,108 @@ For single-configuration build systems:
...
+
+    # remove an existing block
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.pre_blocks.remove("generic_system")
+
+    # modify the template of an existing block
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tmp = tc.pre_blocks["generic_system"].template
```
Comment (danimtb) [inline reference/conanfile/tools/cmake.rst:222]: line not needed?
```suggestion
```

File: reference/conanfile/tools/cmake.rst [hunk 2]
```diff
@@ -192,11 +184,108 @@ For single-configuration build systems:
...
+
+    # modify the context (variables) of an existing block
+    def generate(self):
+        tc = CMakeToolchain(self)
+        generic_block = toolchain.pre_blocks["generic_system"]
+
+        def context(self):
+            assert self  # Your own custom logic here
+            return {"build_type": "SuperRelease"}
+        generic_block.context = types.MethodType(context, generic_block)
```
Comment (danimtb) [inline reference/conanfile/tools/cmake.rst:234]: Where is this `types` coming from? Is this python syntax or something provided by conan?

File: reference/conanfile/tools/cmake.rst [hunk 3]
```diff
@@ -192,11 +184,108 @@ For single-configuration build systems:
...
+    def generate(self):
+        tc = CMakeToolchain(self)
+        # this could go to a python_requires
+        class MyBlock(Block):
+            template = "Hello {{myvar}}!!!"
+
+            def context(self):
+                return {"myvar": "World"}
+
+        tc.pre_blocks["mynewblock"] = MyBlock
```
Comment (danimtb) [inline reference/conanfile/tools/cmake.rst:270]: repeated code, same case just above

File: reference/conanfile/tools/cmake.rst (hunk 1 above)
Comment (memsharded) [reply reference/conanfile/tools/cmake.rst:222]: Sorry, typo in next line

File: reference/conanfile/tools/cmake.rst [hunk 4]
```diff
@@ -192,11 +184,108 @@ For single-configuration build systems:
...
+    # remove an existing block
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.pre_blocks.remove("generic_system")
+
+    # modify the template of an existing block
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tmp = tc.pre_blocks["generic_system"].template
+        new_tmp = temp.replace(...)  # replace, fully replace, append...
```
Comment (memsharded) [inline reference/conanfile/tools/cmake.rst:223]:
```suggestion
        new_tmp = tmp.replace(...)  # replace, fully replace, append...
```

File: reference/conanfile/tools/cmake.rst (hunk 2 above)
Comment (memsharded) [reply reference/conanfile/tools/cmake.rst:234]: Yes, this is native python

File: reference/conanfile/tools/cmake.rst (hunk 3 above)
Comment (memsharded) [inline reference/conanfile/tools/cmake.rst:270]:
```suggestion
```

File: reference/conanfile/tools/cmake.rst (hunk 2 above)
Comment (danimtb) [reply reference/conanfile/tools/cmake.rst:234]: Maybe adding the import above to make it explicit?

**Review by czoido — APPROVED** [review]

**Pull request merged**